{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "9fe8f207",
   "metadata": {},
   "source": [
    "# Convert filtered VCF files to HDF5\n",
    "\n",
    "This document converts `SeqArray`/`SNPRelate` filtered VCF files and converts\n",
    "them to HDF5 objects for use with `Ipyrad` (specifically STRUCTURE)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "14ae0ea4",
   "metadata": {},
   "outputs": [],
   "source": [
    "import ipyrad.analysis as ipa\n",
    "from pathlib import Path"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "28cfda49",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Indexing VCF to HDF5 database file\n",
      "VCF: 2721 SNPs; 148 scaffolds\n",
      "[####################] 100% 0:00:09 | converting VCF to HDF5 \n",
      "HDF5: 2721 SNPs; 1714 linkage group\n",
      "SNP database written to ../results/variant-filtering/ALA-ld-maf.snps.hdf5\n",
      "Indexing VCF to HDF5 database file\n",
      "VCF: 2360 SNPs; 39 scaffolds\n",
      "[####################] 100% 0:00:10 | converting VCF to HDF5 \n",
      "HDF5: 2360 SNPs; 1714 linkage group\n",
      "SNP database written to ../results/variant-filtering/HST-ld-maf.snps.hdf5\n",
      "Indexing VCF to HDF5 database file\n",
      "VCF: 2177 SNPs; 38 scaffolds\n",
      "[####################] 100% 0:00:10 | converting VCF to HDF5 \n",
      "HDF5: 2177 SNPs; 1621 linkage group\n",
      "SNP database written to ../results/variant-filtering/HST-wa-ld-maf.snps.hdf5\n",
      "Indexing VCF to HDF5 database file\n",
      "VCF: 1208 SNPs; 25 scaffolds\n",
      "[####################] 100% 0:00:09 | converting VCF to HDF5 \n",
      "HDF5: 1208 SNPs; 898 linkage group\n",
      "SNP database written to ../results/variant-filtering/HMA-ld-maf.snps.hdf5\n"
     ]
    }
   ],
   "source": [
    "dirpath = Path(\"../results/variant-filtering\")  # Replace with your actual directory\n",
    "\n",
    "for vcf in dirpath.glob(\"*.vcf.gz\"):\n",
    "    if vcf.is_file():  # Check if the item is a file\n",
    "        basename = vcf.stem.split(\".\")[0]\n",
    "        dirname = vcf.parent\n",
    "\n",
    "        # Unfiltered Ipyrad output\n",
    "        converter = ipa.vcf_to_hdf5(\n",
    "            name=f\"{basename}\",\n",
    "            data=vcf.as_posix(),\n",
    "            workdir=dirname,\n",
    "            ld_block_size=50000\n",
    "        )\n",
    "\n",
    "        # run the converter\n",
    "        converter.run(force=True)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
